### ISSUE TITLE
Find and Filter Data From Others Tabel But Need To initialize Model Name (string)

### ISSUE THREAD

**Issue by halimkun** [tailflow/laravel-orion#255]
I have data like this in the database.
```
006/A/S-RSIA/090124	1.211.0918	surat/internal	App\Models\RsiaSuratInternal
006/A/S-RSIA/090124	1.221.0918	surat/internal	App\Models\RsiaSuratInternal
006/A/S-RSIA/090124	1.214.0918	surat/internal	App\Models\RsiaSuratInternal
013/A/S-RSIA/220124	1.209.0918	surat/internal	App\Models\RsiaSuratInternal
013/A/S-RSIA/220124	1.210.0918	surat/internal	App\Models\RsiaSuratInternal
020/A/S-RSIA/290124	2.475.0123	surat/internal	App\Models\RsiaSuratInternal
020/A/S-RSIA/290124	2.475.0123	surat/internal	App\Models\RsiaSuratInternal
023/A/S-RSIA/050224	3.702.0317	surat/internal	App\Models\RsiaSuratInternal
023/A/S-RSIA/050224	3.722.0317	surat/internal	App\Models\RsiaSuratInternal
023/A/S-RSIA/050224	3.752.0317	surat/internal	App\Models\RsiaSuratInternal
033/A/S-RSIA/150224	2.301.0502	surat/internal	App\Models\RsiaSuratInternal
036/A/S-RSIA/170224	1.223.0523	surat/internal	App\Models\RsiaSuratInternal
036/A/S-RSIA/170224	1.223.0533	surat/internal	App\Models\RsiaSuratInternal
036/A/S-RSIA/170224	1.223.0513	surat/internal	App\Models\RsiaSuratInternal
036/A/S-RSIA/170224	1.223.0512	surat/internal	App\Models\RsiaSuratInternal
040/A/S-RSIA/010324	2.310.0113	surat/internal	App\Models\RsiaSuratInternal
041/A/S-RSIA/040324	2.493	        surat/internal	App\Models\RsiaSuratInternal
047/A/S-RSIA/130324	1.223.0523	surat/internal	App\Models\RsiaSuratInternal
049/A/S-RSIA/160324	2.308.1012	surat/internal	App\Models\RsiaSuratInternal
050/A/S-RSIA/180324	1.223.0523	surat/internal	App\Models\RsiaSuratInternal
052/A/S-RSIA/230324	3.910.0102	surat/internal	App\Models\RsiaSuratInternal
057/A/S-RSIA/030424	2.310.0113	surat/internal	App\Models\RsiaSuratInternal
058/A/S-RSIA/030424	3.909.0119	surat/internal	App\Models\RsiaSuratInternal
```

if you see that there is no text that I can search for such as the subject of the letter. where the data `the subject of the letter` **is stored in another table** whose model name I stored in a table in my database. i thought that storing the model name would make it easier to search for data in the future but I found a dead end when trying to apply data search and filter because it needs to ___initialize the model___ first. 

menapa tidak merelasikan tabelnya ke tabel sumber ? 
tabal ini adalah tabel **penerima** dan tabel sumbernya **sesuai dengan nama model yang disimpan**. yang mana tabel penerima ini bisa mendapatkan isi dari `bebrapa tabel sumber` sesuai dengan `identifier unik pada fileds pertama`. 

I solved it with a code like this. in the collection.
```php
        return $this->collection->map(function ($item) use ($request) {
            $keywords = $request->get('search');

            // Membuat instance model dari item
            $model = new $item->model;

            // Melakukan pencarian menggunakan kata kunci pada model
            $query = $model->where('no_surat', $item->no_surat);
            if ($keywords) {
                $query = $query->where(function ($q) use ($keywords) {
                    $q->where('perihal', 'like', '%' . $keywords['value'] . '%');
                });
            }
            $undangan = $query->first();

            // Mengambil jumlah penerima dan tipe undangan berdasarkan no_surat
            $penerimaCount = RsiaPenerimaUndangan::where('no_surat', $item->no_surat)->count();
            $undanganTipe = RsiaPenerimaUndangan::where('no_surat', $item->no_surat)->first();

            // Verifikasi apakah $undangan adalah array atau objek
            if (is_array($undangan)) {
                $undangan = (object) $undangan;
            }

            return [
                'no_surat' => $item->no_surat,
                'tipe' => $undanganTipe ? $undanganTipe->tipe : null,
                'penerima_count' => $penerimaCount,
                'undangan' => $undangan,
            ];
        })->filter(function ($item) {
            return $item['undangan'] !== null;
        })->values()
```

but do you know what happens? 
when the data I'm looking for is not on the first page but is on the second page, the first page will still be displayed as a response with empty data and the second page will display the data I'm looking for. 
> This happens because I do the search right after the pagination is created. So the filter is only applied to the pagination data that is loaded by me, the data on the pagination that has not been loaded will be applied when accessed.

So basically, can I do a search with my case? 
where I need to init the table first?

thank you for your attention.

**Comment by alexzarbn** [comment]
Hi @halimkun,

I'm afraid such use case is not possible at the moment.

**Comment by halimkun** [comment]
I've reopened this issue to share a solution that I've successfully implemented for handling complex data retrieval and search challenges involving multiple related models in Laravel Orion. Here's a summary of what I've done:

### Problem Statement
I had data stored in a table `rsia_penerima_undangan `where each row referenced a specific model (`RsiaSuratInternal`, etc.) through a polymorphic relation. The goal was to search for records based on attributes stored in these related models (`perihal `in `RsiaSuratInternal`, etc., for example).

### Solution Implemented
1. Polymorphic Relation in `RsiaPenerimaUndangan` Model:
    ```php
    public function relatedModel(): \Illuminate\Database\Eloquent\Relations\MorphTo
    {
        return $this->morphTo('no_surat', 'model', 'no_surat', 'no_surat');
    }
    ```
2. Query Scope for Search in `RsiaPenerimaUndangan` Model
    ```php
    public function scopeSearchByRelatedModel($query, $searchTerm)
    {
        return $query->whereHas('relatedModel', function ($query) use ($searchTerm) {
            $query->where('perihal', 'like', "%{$searchTerm}%");
        });
    }
    ```
3. Controller Method to Handle Index Fetching
    ```php
    // PenerimaUndanganController.php
    protected function runIndexFetchQuery(Request $request, \Illuminate\Database\Eloquent\Builder $query, int $paginationLimit)
    {
        $searchTerm = $request->input('search.value', '');
        $query = $query->searchByRelatedModel($searchTerm);
        
        return $this->shouldPaginate($request, $paginationLimit) ? $query->paginate($paginationLimit) : $query->get();
    }
    
    public function exposedScopes(): array
    {
        return ['scopeSearchByRelatedModel'];
    }
    ```

### How to Test
You can test this solution by making a POST request to your search endpoint (/api/penerima-undangan/search), sending a JSON payload like this:
```json
{
    "search": {
        "value": "your_search_term_here"
    }
}
```

### Additional Notes
I'm not entirely certain if this approach will work universally for all scenarios, or if there are specific prerequisites such as storing model names in the database, among other considerations. Therefore, please manage expectations accordingly. Also, I haven't yet tested if this solution effectively filters data across the source tables. I'd greatly appreciate any feedback or alternative solutions regarding this filtering aspect.

I hope this solution proves helpful to others facing similar challenges. Please feel free to provide feedback or ask any questions. Thank you!
